src/licenses/LicenseModel.cpp

```diff
@@ -0,0 +1,184 @@
+#include "LicenseModel.h"
+
+#include <QDirIterator>
+
+static const char *const platformRootPath = ":/licenses";
+static const char *const qtRootPath = ":/licenses/qt";
+static const char *const toolkitRootPath = ":/licenses/qt/licenses";
+static const char *const webviewRootPath = ":/licenses/qt/licenses/qtwebengine/src/3rdparty/chromium";
+
+namespace licenses {
+
+  LicenseModel::LicenseModel(QObject *parent) : QAbstractItemModel(parent), m_dir(platformRootPath) {
+    populate();
+  }
+
+  LicenseModel::~LicenseModel() = default;
+
+  void LicenseModel::populate() {
+    QList<QVariant> root_categories;
+    QVector<QVariant> rootData = {QString("Licenses"), QString(platformRootPath)};
+    rootItem = std::make_unique<LicenseItem>(rootData, root_categories);
+
+    QHash<QString, LicenseItem *> map;
+    map[rootItem->path()] = rootItem.get();
+
+    QDirIterator it(m_dir, QDirIterator::Subdirectories);
+
+    while (it.hasNext()) {
+      QString path = it.next();
+
+      QList<QVariant> categories;
+      categories.append(QVariant(static_cast<int>(LicenseCategory::All)));
+
+      if (path.startsWith(webviewRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::WebView)));
+
+      if (path.startsWith(toolkitRootPath) && !path.startsWith(webviewRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::Toolkit)));
+
+      if (path.startsWith(platformRootPath) && !path.startsWith(qtRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::Platform)));
+
+      const QFileInfo &fileInfo = it.fileInfo();
+      LicenseItem *parent = map[fileInfo.dir().path()];
+
+      QVector<QVariant> child_data = {fileInfo.fileName(), fileInfo.absoluteFilePath()};
+      auto *child = new LicenseItem(child_data, categories, parent);
+      parent->appendChild(child);
+
+      if (fileInfo.isDir())
+        map[child->path()] = child;
+    }
+  }
+
+  QModelIndex LicenseModel::index(int row, int column, const QModelIndex &parent) const {
+    if (!hasIndex(row, column, parent))
```

Comment:
does hasIndex do what you want if parent is invalid?

I have no idea, but it looks like you replace parent with rootItem when parent is invalid, but you don't do that for hasIndex(), so I assume you know what you are doing, but I can only assume, I can't be sure.

Reply:
hasIndex only checks if row and column are in range and it does that by calling the pure virtual columnCount and rowCount. columnCount just returns 1 for me and rowCount checks if parent is valid.

Reply:
Perhaps the "right thing" would be for columnCount to return 0 for invalid parent?

Reply:
Looks like columnCount is ok according to the documentation:
https://doc.qt.io/qt-5/qabstractitemmodel.html#columnCount